pkg/sqlmodel/where_handle.go

```diff
@@ -0,0 +1,153 @@
+	UniqueIdxs []*model.IndexInfo
+}
+
+// GetWhereHandle calculates a WhereHandle by source/target TableInfo's indices,
+// columns and state. Other component can cache the result.
+func GetWhereHandle(source, target *model.TableInfo) *WhereHandle {
+	ret := WhereHandle{}
+	indices := make([]*model.IndexInfo, 0, len(target.Indices)+1)
+	indices = append(indices, target.Indices...)
+	if idx := getPKIsHandleIdx(target); target.PKIsHandle && idx != nil {
+		indices = append(indices, idx)
+	}
+
+	for _, idx := range indices {
+		if !idx.Unique {
+			continue
+		}
+		// when the tableInfo is from CDC, it may contain some index that is
+		// creating.
+		if idx.State != model.StatePublic {
+			continue
+		}
+
+		rewritten := rewriteColsOffset(idx, source)
+		if rewritten == nil {
+			continue
+		}
+		ret.UniqueIdxs = append(ret.UniqueIdxs, rewritten)
+
+		if rewritten.Primary {
+			// PK is prior to UNIQUE NOT NULL for better performance
+			ret.UniqueNotNullIdx = rewritten
+			continue
+		}
+		// use downstream columns to check NOT NULL constraint
+		if ret.UniqueNotNullIdx == nil && allColsNotNull(idx, target.Columns) {
+			// TODO: considering cardinality
+			ret.UniqueNotNullIdx = rewritten
+			continue
+		}
+	}
+	return &ret
+}
+
+// rewriteColsOffset rewrites index columns offset to those from source table.
+// Returns nil when any column does not represent in source.
+func rewriteColsOffset(index *model.IndexInfo, source *model.TableInfo) *model.IndexInfo {
+	if index == nil || source == nil {
+		return nil
+	}
+
+	columns := make([]*model.IndexColumn, 0, len(index.Columns))
+	for _, key := range index.Columns {
+		sourceColumn := model.FindColumnInfo(source.Columns, key.Name.L)
+		if sourceColumn == nil {
+			return nil
+		}
+		column := &model.IndexColumn{
+			Name:   key.Name,
+			Offset: sourceColumn.Offset,
+			Length: key.Length,
+		}
+		columns = append(columns, column)
+	}
+	clone := *index
+	clone.Columns = columns
+	return &clone
+}
+
+func getPKIsHandleIdx(ti *model.TableInfo) *model.IndexInfo {
+	if pk := ti.GetPkColInfo(); pk != nil {
+		return &model.IndexInfo{
+			Table:   ti.Name,
+			Unique:  true,
+			Primary: true,
+			State:   model.StatePublic,
+			Tp:      model.IndexTypeBtree,
+			Columns: []*model.IndexColumn{{
+				Name:   pk.Name,
+				Offset: pk.Offset,
+				Length: types.UnspecifiedLength,
+			}},
+		}
+	}
+	return nil
+}
+
+func allColsNotNull(idx *model.IndexInfo, cols []*model.ColumnInfo) bool {
+	for _, idxCol := range idx.Columns {
+		col := cols[idxCol.Offset]
+		if !mysql.HasNotNullFlag(col.Flag) {
+			return false
+		}
+	}
+	return true
+}
+
+// getWhereIdxByData returns the index that is identical to a row change, it
+// may be
+// - a PK, or
+// - an UNIQUE index whose columns are all NOT NULL, or
+// - an UNIQUE index and the data is all NOT NULL.
+// For the last case, last used index is swapped to front.
+func (h *WhereHandle) getWhereIdxByData(data []interface{}) *model.IndexInfo {
+	if h == nil {
```

Comment:
seems h never equal to nil?

Reply:
Just to improve robustness because WhereHandle is public

Reply:
maybe log something, should be a bug in most usage.